File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        std::queue<TreeNode*> q;
        q.push(root);
        std::vector<std::vector<int>> res;
        while (!q.empty()) {
            int q_size = q.size();
            std::vector<int> curr_level;
            curr_level.reserve(q_size); // This prevents multiple reallocations of the vector as it grows inside the loop
            for (int i = 0; i < q_size; i++) {
                TreeNode* node = q.front();
                q.pop();
                curr_level.push_back(node->val);
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            res.push_back(curr_level);
        }
        return res;     
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def kth(root):
            if not root: return -1

            var = self.kthSmallest(root.left, self.k)
            if var != -1:
                return var
            if self.k == 1:
                return root.val
            self.k -= 1
            return self.kthSmallest(root.right, self.k)
        self.k = k
        return kth(root)    